flink/src/main/java/org/apache/iceberg/flink/CatalogLoader.java

```diff
@@ -45,12 +45,22 @@
    */
   Catalog loadCatalog();
 
-  static CatalogLoader hadoop(String name, Configuration hadoopConf, String warehouseLocation) {
-    return new HadoopCatalogLoader(name, hadoopConf, warehouseLocation);
+  static CatalogLoader hadoop(
+      String name,
+      Configuration hadoopConf,
+      String warehouseLocation,
+      Map<String, String> properties) {
+    return new HadoopCatalogLoader(name, hadoopConf, warehouseLocation, properties);
```

Comment:
For Flink, I think we should pass the warehouse location and other config through properties. Right now, we pull it out in `FlinkCatalogFactory`, but it doesn't make much sense to pull out only some properties.

How about leaving the current `CatalogLoader.hadoop` and `CatalogLoader.hive` as they are and adding variants like `CatalogLoader.hadoop(String name, Map<String, String> properties, Configuration conf)`? Then we can pull the properties out using standard config properties that we put in `CatalogProperties` in the loader classes.

FYI @JingsongLi and @openinx: we're improving how we configure catalogs and allowing some parts to be loaded dynamically. The main change here is passing properties to the catalog as a map.

Reply:
I only see the `CatalogLoader.hadoop` and `CatalogLoader.hive` used in a single place `FlinkCatalogFactory.createCatalogLoader`, that is why I directly changed the class signature. I don't know if there is any benefit in keeping the old ones. 

Changing the signature to `CatalogLoader.hadoop(String name, Map<String, String> properties, Configuration conf)` looks like a good idea that simplifies the code, let me do that first.

(speaking of this, I should also add null check for the properties map and also give a fixed serialization id for those classes because Flink serializes the catalog loader)

Reply:
You're right. Since these are package-private, I don't think we need to maintain them. I was thinking that people would call these from the public API to configure the source and sink builders, but these aren't public.

We may still want to keep them to avoid changing lots of files in this PR, and I still think it is a good idea to pull out the config here, rather than in `FlinkCatalogFactory` because we want to move toward catalogs interpreting their own property maps.

Reply:
> also give a fixed serialization id for those classes because Flink serializes the catalog loader

We know that serialization across Iceberg versions may be a problem, but I'm not sure that we want to introduce a serialization id. In general, we avoid this because we _want_ serialization to fail if there are multiple library versions. Java serialization isn't something that we want to use for compatibility across versions. The cases where we have multiple library versions are rare, and we want to design something for a rolling update.